### ISSUE TITLE
Representation Colour Help

### ISSUE THREAD

**Issue by DustBytes** [nglviewer/ngl#855]
Hi.

I am trying to make a viewer where user can view certain residues as of different colors based on the groups.

What I tried is  : https://jsfiddle.net/oL7jvhew/

Try moving cursor on the right side panel list. Wait for few seconds.
Issue: It's slow. It's really slow. Those few seconds it takes to change the color is the problem.
For smaller proteins it works OK. But for bigger structures it becomes impractical. 
How to improve the performance?

**Comment by fredludlow** [comment]
It's slightly more performant if you don't recreate the representations every time:
https://jsfiddle.net/4q8cyf10/2/

But it's still not great, which is to do with the way the selection engine works. A bit of a workaround is here:
https://jsfiddle.net/4q8cyf10/6/

Basically, the selection code tests each atom against the whole of your expression each time, which, given its structure (lots of chain, residue ID stuff) is slow, we can make a more efficient representation by getting a BitArray (getAtomSet) and then calling `toSeleString()` on it - this is a string that looks like `"@1,2,3,110,12121,..."`

My apologies that this is very hacked to your specific case, but the relevant snippet from the fiddle linked above is:

```javascript
var colourScheme = "";
            var representation = "";
            
            var sele1 = new NGL.Selection(colorData[0].join(''))
            var sele2 = new NGL.Selection(colorData[1].join(''))
            
            var comp = stage.compList[0]
            var struc = comp.structure
            var atomSeleExpr1 = struc.getAtomSet(sele1).toSeleString()
            var atomSeleExpr2 = struc.getAtomSet(sele2).toSeleString()
            
            
            
            var sid = generateSchemeID([[atomSeleExpr1], [atomSeleExpr2]])

            stage.getRepresentationsByName("polymer").setParameters({
            	// colorScheme: generateSchemeID(colorData)
              colorScheme: sid
              // colorScheme: schemes[colorSchemeIndex++ % 2]
            });
```


Two solutions to this (in terms of fixing NGL code):
1) Do the AtomSet creation inside `SelectionColormaker.atomColor` (atomSets are cached on `Structure` so would only get called once)
2) Allow passing in `BitArrays` rather than selection expressions

**Comment by DustBytes** [comment]
It works. Thank You.
It will take me some time to understand how it is working.

Two more things I am struggling with, related to this.
Issue  #822
https://github.com/nglviewer/ngl/issues/822

I was not able to explain the issue properly there.
1) Simply, have  a look at the fiddle you posted. The whole structure is of white/gray color except the red, blue groups. I need to put colors on the chains also, so that it's easy to understand where two chains are interacting. I tried manually adding the residue list to the "generateSchemeID" after red, blue groups but that even the first loading is slow. If making them of light colors or of little transparency, it will be better looking.

2) Second, there are domains in proteins. I need to add boundaries of those domains. Let's say if N residues are forming a domain in chain X, make a dark outline forming an area of boundary residues  Or color them of some other shade of the same colour as chain.
These red,blue interactions happen chain wise and domain wise. ( later need to add trasparency in red, blue also )


I made a different server where on mouehover 2 residues change to ball+stick on cartoon representation, telling the user where the bonds exist. I will try to implement similar logic there to increase the speed.

These kind of features, where we can show certain interactions, residues, representations is very common. JSmol got some scripts which can be implemented directly and it's fast. That's how I was trying to implement same in ngl.
Hope these kind of stuff is implemented in mol*.  


Thank You
